cl/alerts/tasks.py

```diff
@@ -28,117 +30,142 @@ def enqueue_docket_alert(d_pk: int) -> bool:
     return create_redis_semaphore("ALERTS", key, ttl=60 * 10)
 
 
+def get_docket_alert_recipients(
+    d_pk: int,
+    recap_email_recipients: list[str] = [],
+) -> tuple[
+    list[tuple[str, str]],
+    list[tuple[str, str]],
+    list[tuple[str, str]],
+    list[str],
+]:
+    """Get the notification's recipients for a docket alert.
+    :param d_pk: Docket primary key
+    :param recap_email_recipients: List of @recap.email addresses to send the
+    notification to.
+    :return: A Tuple of lists of tuples (email_address, secret_key) and a list
+    of @recap.email addresses that don't belong to any user if any.
+    """
+
+    # Lists of tuples (email_address, docket_alert.secret_key)
+    docket_alert_recipients_list = []
+    recap_email_first_time_auto_subscribe_list = []
+    recap_email_first_time_no_auto_subscribe_list = []
+    # List of @recap.email addresses that don't belong to any user
+    recap_email_user_does_not_exist_list = []
+
+    # First, get current docket alert recipients to avoid duplicate alerts
+    docket_alerts_current_subscribers = DocketAlert.objects.filter(
+        docket_id=d_pk, alert_type=DocketAlert.SUBSCRIPTION
+    )
+    for da in docket_alerts_current_subscribers:
+        recipient_tuple = (da.user.email, da.secret_key)
```

Comment:
We need to remember to focus on queries in this code. So if you need to get `da.user.email`, remember to use a `select_related` up above.

Reply:
Perfect, I added it here and in a different query below where it's also needed to access a related object field.